### PR TITLE
placeholder from Speechllm selene to main

### PR DESCRIPTION
**Pull request by zhehuaichen**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 1]
```diff
@@ -159,7 +169,7 @@ def _metrics_require_string2category_map(self):
     def setup(self, stage=None):
         # NOTE: super().__init__ will try and setup train/val/test datasets, but we sidestep this using a if self._train_ds is not None condition
         # We then set things up for real only once setup() of this class is called.
-        resume_checkpoint_path = self.trainer._checkpoint_connector.resume_from_checkpoint_fit_path
+        resume_checkpoint_path = self.trainer.ckpt_path
```
Comment (zhehuaichen) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:172]: hi @stevehuang52 do you  remember whether and why this change is necessary ?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 2]
```diff
@@ -638,19 +732,36 @@ def cast_for_metric(self, pred, label, metric_name, class_labels=None, labels_ar
...
-            app_state = AppState()
-            _reconfigure_microbatch_calculator(
-                rank=app_state.global_rank,
-                rampup_batch_size=None,
-                global_batch_size=global_batch_per_gpu * parallel_state.get_data_parallel_world_size(),
-                micro_batch_size=global_batch_per_gpu,
-                data_parallel_size=parallel_state.get_data_parallel_world_size(),
-            )
-        return batch
+    def _reconfigure_and_process_inference_batch(self, batch, data_cfg):
```
Comment (zhehuaichen) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:735]: @stevehuang52 do you remember whether and why this change is necessary?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 3]
```diff
@@ -708,32 +819,16 @@ def setup_training_dataloader(self):
...
     def setup_eval_dataloader(self, datasets, data_cfg):
         dataloaders = []
+        if not isinstance(datasets, list):
+            return self.build_data_loader(dataset=datasets, data_cfg=data_cfg, consumed_samples=0,)
         for dataset in datasets:
             eval_dl = self.build_data_loader(dataset=dataset, data_cfg=data_cfg, consumed_samples=0,)
             dataloaders.append(eval_dl)
         return dataloaders
 
-    def _reset_activation_checkpointing_args(self):
```
Comment (zhehuaichen) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:716]: @stevehuang52 do you remember whether and why this change is necessary?

**Review by zhehuaichen — COMMENTED** [review]
left some items to run down

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py (hunk 1 above)
Comment (stevehuanghe) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:172]: I think we need to first merge main into our branch, resolve the conflicts, then merge our branch into main, since the main branch has changed a lot since we last pull

Comment (stevehuanghe) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:172]: @zhehuaichen just noticed that you're merging into your own main branch, not the latest NeMo main. In this case, you can just keep all changes in the `Megatron_gpt_sft_model.py` for now. The original PR where I copied the changes from is this one [6915](https://github.com/NVIDIA/NeMo/pull/6915), but I made a few changes to it to keep it work with PTL 1.x instead of 2.x used in NeMo main.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 4]
```diff
@@ -236,6 +246,16 @@ def _build_dataset(self, data_cfg, is_train=True):
         else:
             num_train_samples_per_dataset = [[None]] * len(data_cfg.file_names)
 
+        # Check dataset max_seq_legnth and max_position_embeddings size
```
[the remark this answers is not in the document]
Comment (stevehuanghe) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:249]: https://github.com/NVIDIA/NeMo/pull/6915

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py (hunk 2 above)
Comment (stevehuanghe) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:735]: it's from the source PR https://github.com/NVIDIA/NeMo/pull/6915

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py (hunk 3 above)
Comment (stevehuanghe) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:716]: it's from the source PR https://github.com/NVIDIA/NeMo/pull/6915

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 5]
```diff
@@ -359,179 +384,206 @@ def fwd_bwd_step(self, dataloader_iter, batch_idx, forward_only):
...
-            )
-            _ = metric(pred, label)
-
-        return {
-            'loss': loss,
-            'preds': preds_text,
-            'labels': labels_text,
-            'inputs': input_text,
-        }
+        # if mode == 'validation':
```
Comment (stevehuanghe) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:436]: these lines are commented out since there's no `self.validation_step_outputs` in the model, maybe PTL2.0 has it

Comment (stevehuanghe) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:436]: these lines are commented out since there's no `self.validation_step_outputs` in the model, maybe PTL2.0 has it

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 6]
```diff
@@ -541,46 +593,88 @@ def inference_epoch_end(self, outputs, mode, data_cfg):
             averaged_metric = 0.0 if monitor_mode == 'max' else 1e5
 
         if mode == 'validation':
-            self.log("validation_loss", averaged_loss)
+            self.log("validation_loss", averaged_loss, batch_size=1, sync_dist=True)
```
Comment (stevehuanghe) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:596]: may not be necessary, I added `batch_size=1, sync_dist=True` just because other places have them. it's ok to keep the original code.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 7]
```diff
@@ -756,37 +852,21 @@ def on_test_epoch_start(self):
         )
         return super().on_test_epoch_start()
 
-    def on_test_epoch_end(self):
-        self.on_inference_epoch_end(self.cfg.data.test_ds)
-        return super().on_test_epoch_end()
+    # def on_test_epoch_end(self):
```
Comment (stevehuanghe) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:855]: commented out since they are not actually called in PTL1.8
